Add reader for GERB high-resolution HDF5 files

pdebuyl opened the pull request:
<!-- Describe what your PR does, and why -->
<!-- For works in progress choose "Create draft pull request" from the drop-down green button. -->

 - [ ] Closes #xxxx <!-- remove if there is no corresponding issue, which should only be the case for minor changes -->
 - [ ] Tests added <!-- for all bug fixes or enhancements -->
 - [ ] Fully documented <!-- remove if this change should not be visible to users, e.g., if it is an internal clean-up, or if this is part of a larger project that will be documented later -->
 - [ ] Add your name to `AUTHORS.md` if not there already

Hi satpy,

I propose a reader for the L2 HR product of the GERB instrument. The data documentation is available here: https://gerb.oma.be/doku.php?id=documentation

I put three sample files here: https://gerb.oma.be/public/pdebuyl/sample_data/

I would be glad to see this included in satpy.

Quick background: GERB stands for "Geostationary Earth Radiation Budget". The project, for which I work, produces top of atmosphere fluxes in the reflected short wave domain and in the outgoing long wave domain (thermal).

The images are 1237 by 1237 pixels, which is the 9km (nadir) version of the SEVIRI grid. The instrument flies on MSG satellites.

I am happy to answer to any question of course and to improve my code to reach the requirements of satpy. Example of use:

```
import satpy
scene = satpy.Scene(reader="gerb_l2_hr_h5", filenames=["G1_SEV2_L20_HR_SOL_TH_20120621_101500_ED01.hdf"])
scene.load(['Thermal Flux', 'Solar Flux'])
scene.show('Thermal Flux')
scene.show('Solar Flux')
```

djhoese commented on satpy/readers/gerb_l2_hr_h5.py:
```diff
@@ -0,0 +1,110 @@
+        ds_real = ds[...]*ds.attrs['Quantisation Factor']
+    else:
+        ds_real = ds[...]*1.
+    ds_min = ds[...].min()
+    if ds_min < 0:
+        mask = ds == ds_min
+        ds_real[mask] = np.nan
```
I'll mention other ideas for dask-ification in my main review comment, but wanted to point out that this is very inefficient and if/when it is done with dask arrays it will be even more inefficient. It was hard for me to understand the product guide regarding fill values, but it looks like the fill or "Error" value is always the minimum value of the in-file data type. Does that sound right?

In that case, can we hardcode this fill value in the YAML or look it up using something like `np.iinfo(ds.dtype).min`. Or are there attributes in the file that tell us this information? That way we don't have to look at every pixel just to determine the minimum value that we use for masking. Also, why does the minimum value have to be `< 0` to be used for masking?

djhoese commented on satpy/readers/gerb_l2_hr_h5.py:
```diff
@@ -0,0 +1,110 @@
+    @property
+    def end_time(self):
+        """Get end time."""
+        return self.start_time + timedelta(minutes=14, seconds=59)
```
Normally this `end_time` is the nominal time of the data and is usually pretty "human friendly". I'm guessing the repeat cycle of the instrument is 15 minutes? In that case could we make this 15 minutes instead of 14:59? Or in some cases we just make this equal to `start_time` if we can't get it from the file data.

simonrp84 replied:
Would making the end time exactly 15 mins not present problems as it'll overlap with the start time of the next scan, though?

pdebuyl replied:
@simonrp84 that was my motivation (plus the fact that some other readers seem to do that)

djhoese replied:
@pdebuyl What other readers? I tried to find it when I made my comment, but I was pretty sure @mraspaud and I defined `end_time`s in Satpy as being exclusive. So start/end times being every 15 minutes would be fine because any downstream application should be interpreting the end time as happening the instant before the time specified.

pdebuyl replied:
By chance, I had a look at the `hsaf_h5` reader (for inspiration as it is another HDF5 reader) and that one use 23 hours 59 minutes, hence my confusion. I'll put 15 minutes.

djhoese replied:
It looks like that was a suggestion from @pnuu here:

https://github.com/pytroll/satpy/pull/2285#discussion_r1036763660

Perhaps we need to discuss this further. I know the FCI folks and I had trouble in our SIFT application when the nice human-friendly times weren't used because it was awkward to represent the blocks of time in a GUI...but maybe that was more about using real observation/scanning times rather than off-by-one inclusive times versus exclusive times.

djhoese commented on satpy/etc/readers/gerb_l2_hr_h5.yaml:
```diff
@@ -0,0 +1,31 @@
+  Thermal_Flux:
+    name: Thermal Flux
+    sensor: gerb
+    wavelength: [4., 22., 40.]
```
I'm surprised these datasets have an associated wavelength. If this information isn't generally used or doesn't make sense for other products you've wanted to add it is optional and doesn't need to be specified.

pdebuyl replied:
It is indeed not necessary, I'll remove them.

pdebuyl commented:
Hi @djhoese thanks a lot for taking the time to review the reader. I'll proceed step by step, starting with linter issues.

pdebuyl commented on satpy/readers/gerb_l2_hr_h5.py:
```diff
@@ -0,0 +1,110 @@
+        elif abs(self.ssp_lon - 45.5) < 1e-6:
+            return get_area_def("msg_seviri_iodc_9km")
+        else:
+            raise ValueError
```
OK, I didn't know it was preferred. I actually started using a proj config line, would this be better? There are several approaches in the existing readers.

In any case, the grid for GERB is fixed for the lifetime of the instrument, so the only variable parameter is the SSP longitude.

djhoese replied:
> the grid for GERB is fixed for the lifetime of the instrument

If I had a dollar for every time I've heard something about an instrument isn't supposed to change...

I guess when I say preferred I mean it is my preference. Normally the data files provide their own geolocation information. Otherwise I would consider the files at fault. You're putting the data in the files, describe it. In some readers we've had to hardcode geolocation/area information but only as a last resort. I guess what you've done here is the best we're going to get.

pdebuyl replied:
In principle, we can provide a HDF5 file, and we do when releasing a full dataset. For users wishing to download a short time series, this will be problematic however :-/

djhoese replied:
In the NetCDF CF realm of data files providing the projection information shouldn't be more than 10s or maybe 100 bytes (ex. `grid_mapping` variables). There would also then need to be `x` and `y` variables for the pixel coordinates so there is some extra weight to adding that information.

What you've done here is fine.

pdebuyl commented on satpy/etc/readers/gerb_l2_hr_h5.yaml:
```diff
@@ -0,0 +1,31 @@
+reader:
+  name: gerb_l2_hr_h5
```
In practice, the L2 HR (9km) product is the one we recommend to users. We have a so-called ARG L1.5 (50 km) product and a BARG L2 (50km). Those are the official products of the GERB instrument (RAL Space, Imperial College and the Royal Meteorological Institute of Belgium produce the data on behalf of EUMETSAT).

There might be an issue if/when we produce "edition 2" data in the future that might be in NetCDF.

I think that `gerb_l2_h5` should be fine though :-)

pdebuyl commented:
Hey @djhoese 

I addressed several issues but not all. What remains, is the following:

1. The area definition: if the seviri grids remain in place, which I guess will be the case, is it an issue to use those are definitions?
2. I prefer to keep `gerb_l2_hr_h5` for the name if possible.
3. I removed the call to `.min()` I hope the current version is cleaner.
4. Regarding the choice between 14 minutes 59s and 15 minutes, I don't know what is preferred.

There is timing data in the files as well, I could parse that if it is better for Satpy. I attach an example of the data below.

```
HDF5 "G4_SEV4_L20_HR_SOL_TH_20190808_201500_V010.hdf" {
DATASET "/Times/Time (per row)" {
   DATATYPE  H5T_STRING {
      STRSIZE 22;
      STRPAD H5T_STR_NULLTERM;
      CSET H5T_CSET_ASCII;
      CTYPE H5T_C_S1;
   }
   DATASPACE  SIMPLE { ( 1237 ) / ( 1237 ) }
   DATA {
   (0): "20190808 20:27:39.400", "20190808 20:27:38.799",
   (2): "20190808 20:27:38.198", "20190808 20:27:37.598",
   (4): "20190808 20:27:36.997", "20190808 20:27:36.396",
   (6): "20190808 20:27:35.796", "20190808 20:27:35.195",
   (8): "20190808 20:27:34.594", "20190808 20:27:33.994",
...
   (1234): "20190808 20:15:18.201", "20190808 20:15:17.600",
   (1236): "20190808 20:15:17.000"
   }
}
}
```

djhoese commented on satpy/readers/gerb_l2_hr_h5.py:
```diff
@@ -0,0 +1,108 @@
+        ds = ds*ds_attrs['Quantisation Factor']
+    else:
+        ds = ds*1.
+    ds[fill_mask] = np.nan
```
This works for me. In the future when this is more dask based (maybe by using the helper HDF5 file handler) I think dask would like a `data_arr = data_arr.where(fill_mask)` but maybe we worry about that later. I think in the long run what I'm saying is that this reader should do as much in dask-land as possible. I'm not sure we have a hard requirement about when and where dask is used, but typically we do it as much as possible otherwise performance is generally wasted.

pdebuyl replied:
Done.

djhoese commented:
1. SEVIRI area definitions is fine.
2. Reader name is fine. I'd prefer to drop the `_hr_` but it's your reader so you probably know better than I do.
3. Yes, cleaner. I made a comment about future dask-friendly stuff.
4. For time, you could parse the file, but for such a simple property maybe just do 15 minutes. If you or someone else has examples of readers that do the equivalent of 14:59 then we can open the discussion to other pytroll developers in slack. I'm 95% sure that it should be 15 minutes though.

djhoese commented:
I think overall I'd like to see the use of the HDF5 utility file handler base class and tests added.

pdebuyl commented:
Hi David, at this point I believe the most important is to add tests. Do you have a good example for me to check?

djhoese commented:
> Hi David, at this point I believe the most important is to add tests. Do you have a good example for me to check?

@pdebuyl it depends if you're going to use the HDF5 utility class. You basically have two reasonable options when writing tests for readers:

1. Create a real file on disk that is very similar to the real data files. Write your tests and point to those files. This is a relatively new practice of Satpy developers, but it catches so many more errors since it doesn't require awkward mocking of low-level file reading functionality...but it does require a lot of work for binary file types. HDF5 and NetCDF4 aren't that bad.
2. The utility class has a helper test class as well and can be seen being used in the `viirs_sdr` reader tests and many others. It basically mocks the file opening logic to say "here's a dictionary of variables and metadata that you would have read from the file". This is arguably not as "strong" of a way of doing the tests since we aren't actually testing all of the on-disk file reading, but is nice that all the data is in-memory.

Which ever way you'd like to go with it I can try to point you to more examples.

pdebuyl commented:
I changed to the HDF5 base reader class, that was quite straightforward (at first I feared not being able to access Dataset attributes but that was fine). I'll check the tests, preferably with an actual file as the mock tests confuse me a bit (some extra abstraction there).

pdebuyl commented:
Hello @djhoese I added a test file for the reader. The CI fails, but it is apparently related to a hdf/netcdf package installs under windows, should I care about it?

Regarding the test, I create a real HDF5 containing fake data. I dumped an actual file in a hierarchy of dicts and used it to write code to re-write a file with a similar structure. If this is useful, I can share the code that I wrote for this (it does not take into account HDF5 references and custom dtypes).

djhoese commented:
Merge with upstream `main` and the CI failure related to building the environment should go away (I fixed it and merged it a couple hours ago).

I'll have to take a look at your code, but what you've said seems reasonable.

djhoese commented on satpy/tests/reader_tests/test_gerb_l2_hr_h5.py:
```diff
@@ -0,0 +1,217 @@
+
+
+def test_gerb_solar_flux_dataset(gerb_l2_hr_h5_dummy_file):
+    """Test the GERB L2 HR HDF5 file.
+
+    Load the solar flux component.
+    """
```
We already know we're testing the GERB reader based on the module we're in so you can be a little less detailed in your docstrings. Same goes for the `thermal_flux`.

```suggestion
    """Test loading the solar flux component."""
```

And not that you have to change it, but you could do some fancy stuff with pytest to run both of these tests in the same function:

```python
@pytest.mark.parametrize("name", ["Solar Flux", "Thermal Flux"])
def test_dataset_load(gerb_l2_hr_h5_dummy_file, name):
```

pdebuyl replied:
Applied the later fix with co-authored commit

djhoese commented on satpy/tests/reader_tests/test_gerb_l2_hr_h5.py:
```diff
@@ -0,0 +1,217 @@
+FNAME = "G4_SEV4_L20_HR_SOL_TH_20190606_130000_V000.hdf"
+
+
+@pytest.fixture(scope="session")
+def gerb_l2_hr_h5_dummy_file(tmp_path_factory):
+    """Create a dummy HDF5 file for the GERB L2 HR product."""
+    filename = tmp_path_factory.mktemp("data") / FNAME
```
The code quality checkers are a little upset with how big this function is. How much of this is used in the actual reading code? Is this a complete copy of the files structure? If it isn't all used you could remove it from the test. A lot of it also looks like it could go into a function (copy C_S1, set size, create variable, write, etc). Maybe even a for loop. Thoughts?

pdebuyl replied:
Indeed, this is code generated by a routine that will duplicate some instructions. A lot actually. And very little of the file content is currently used in the code. Basically, the reader loads the most used data and makes sure that the geolocation is correct.

pdebuyl commented:
I rebased with main, refactored a bit the code, removed a lot of the fake file content, and added the datasets `{Solar,Thermal} Radiance` to the reader.

I am ok with it but now I can't plot resampled data. I get weird errors like "'DataArray' object has no attribute 'reshape'" or "'DataArray' object has no attribute 'values'". This seemed to me like the data is not available to the scene anymore so I tested the diff below and it works again.

Note that loading the data as a `xarray.DataArray` worked fine before the rebase from main, so something in the management of the scene's underlying data might have changed.

```.diff
diff --git a/satpy/readers/gerb_l2_hr_h5.py b/satpy/readers/gerb_l2_hr_h5.py
index 4dad36f0e..bf89f3aa3 100644
--- a/satpy/readers/gerb_l2_hr_h5.py
+++ b/satpy/readers/gerb_l2_hr_h5.py
@@ -29,6 +29,7 @@ from datetime import timedelta
 
 import dask.array as da
 import h5py
+import numpy as np
 import xarray as xr
 
 from satpy.readers.hdf5_utils import HDF5FileHandler
@@ -43,7 +44,7 @@ def gerb_get_dataset(hfile, name, ds_info):
 
     The routine takes into account the quantisation factor and fill values.
     """
-    ds = xr.DataArray(hfile[name][...])
+    ds = hfile[name][...]
     ds_attrs = hfile[name].attrs
     ds_fill = ds_info['fill_value']
     fill_mask = ds != ds_fill
@@ -51,7 +52,7 @@ def gerb_get_dataset(hfile, name, ds_info):
         ds = ds*ds_attrs['Quantisation Factor']
     else:
         ds = ds*1.
-    ds = ds.where(fill_mask)
+    ds = np.where(fill_mask, ds, np.nan)
     return ds
```

djhoese commented:
Can you give me a full traceback for one of those errors you're seeing?

pdebuyl commented:
```
(py3.11) pierre@pc:~/RS_DATA/GERB$ python sample_scene.py  --region maspalomas
/opt/py3.11/lib/python3.11/site-packages/dask/array/core.py:3470: UserWarning: Passing an object to dask.array.from_array which is already a Dask collection. This can lead to unexpected behavior.
  warnings.warn(
/opt/py3.11/lib/python3.11/site-packages/dask/array/core.py:3470: UserWarning: Passing an object to dask.array.from_array which is already a Dask collection. This can lead to unexpected behavior.
  warnings.warn(
Traceback (most recent call last):
  File "/home/pierre/RS_DATA/GERB/sample_scene.py", line 21, in <module>
    plt.imshow(local_scene['Thermal Flux'].to_numpy(), transform=crs, extent=crs.bounds, origin='upper', cmap=plt.cm.hot)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/py3.11/lib/python3.11/site-packages/xarray/core/dataarray.py", line 776, in to_numpy
    return self.variable.to_numpy()
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/py3.11/lib/python3.11/site-packages/xarray/core/variable.py", line 1299, in to_numpy
    data, *_ = chunkmanager.compute(data)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/py3.11/lib/python3.11/site-packages/xarray/core/daskmanager.py", line 70, in compute
    return compute(*data, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/py3.11/lib/python3.11/site-packages/dask/base.py", line 628, in compute
    results = schedule(dsk, keys, **kwargs)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/py3.11/lib/python3.11/site-packages/xarray/core/common.py", line 278, in __getattr__
    raise AttributeError(
AttributeError: 'DataArray' object has no attribute 'reshape'. Did you mean: 'shape'?
```

Sample code
```.py
import satpy
import numpy as np
import matplotlib.pyplot as plt
import argparse

parser = argparse.ArgumentParser()
parser.add_argument('--region', default="FD")
parser.add_argument('--save', action="store_true")
args = parser.parse_args()

scene = satpy.Scene(reader="gerb_l2_hr_h5", filenames=["G1_SEV2_L20_HR_SOL_TH_20120621_101500_ED01.hdf"])
scene.load(['Thermal Flux', 'Solar Flux'])

if args.region != "FD":
    local_scene = scene.resample(args.region)
else:
    local_scene = scene

crs = local_scene['Thermal Flux'].attrs['area'].to_cartopy_crs()
ax = plt.axes(projection=crs); ax.coastlines(); ax.gridlines(); ax.set_global()                                                                                                                
plt.imshow(local_scene['Thermal Flux'].to_numpy(), transform=crs, extent=crs.bounds, origin='upper', cmap=plt.cm.hot)
plt.colorbar(orientation='horizontal')
plt.title("GERB Thermal Flux [W/m²]")

if args.save:
    plt.savefig(f'GERB_OLR_201206211015_{args.region}.png')
    plt.clf()
else:
    plt.figure()

ax = plt.axes(projection=crs)
ax.coastlines()
ax.gridlines()
ax.set_global()
plt.imshow(local_scene['Solar Flux'].to_numpy(), transform=crs, extent=crs.bounds, origin='upper', cmap=plt.cm.hot)
plt.colorbar(orientation='horizontal')
plt.title("GERB Solar Flux [W/m²]")

if args.save:
    plt.savefig(f'GERB_RSW_201206211015_{args.region}.png')
else:
    plt.show()
```

invoked as `python sample_scene.py  --region maspalomas`. (any region will do of course). Without the `--region` option there is no error. If you want to check, you can find sample data here: https://gerb.oma.be/public/pdebuyl/sample_data/

djhoese commented on satpy/readers/gerb_l2_hr_h5.py:
```diff
@@ -0,0 +1,105 @@
+
+    The routine takes into account the quantisation factor and fill values.
+    """
+    ds = xr.DataArray(hfile[name][...])
```
This is one reason you're getting the dask errors and warnings you're seeing. The result of `hfile[name]` (I think) should be an xarray.DataArray already. And don't do `[...]` or you'll load it into memory which is not what we want anymore now that the reader is trying to be dask friendly. See my other comment for more info...

pdebuyl replied:
Ok, got it. I am used to h5py and didn't realise that HDF5FileHandler would already take care of that. (+ had a look at the hsaf hdf5 reader that does something similar).

djhoese commented on satpy/readers/gerb_l2_hr_h5.py:
```diff
@@ -0,0 +1,105 @@
+        ds_info['data_time'] = self.start_time
+        ds_info['end_time'] = self.end_time
+
+        data = da.from_array(ds)
+        return xr.DataArray(data, attrs=ds_info, dims=('y', 'x'))
```
`ds` should already be a `DataArray` so doing `da.from_array` is not allowed (you have warnings in your sample script output about it). You also don't need to do `xr.DataArray(data, ...)` either but could instead rename the dimensions with `ds = ds.rename({"some_dim_x": "x", "some_dim_y": "y"{)` and `ds.attrs.update(ds_info)`.

pdebuyl replied:
Using the proper dataset directly, I only need the timing attributes added via `ds.attrs.update`.

pdebuyl commented:
Hi @djhoese thanks for the hand-holding :-)

This seems much cleaner now from the xarray dataset handling. Tests + example program pass on my side, I'll be happy to resolve any further issue if necessary.

djhoese reviewed (commented):
The test file creation looks a lot better. There's still a lot of duplicate code there, but maybe I'm not looking close enough at the details. A function that takes the variable name, the array dtype, and the quantisation factor (always f64) could replace 4 or more variable creations, right? Even the ones that need Offset, they could call this additional helper function and then add Offset afterward.

If you're out of time and won't be able to get to this let me know.

djhoese commented on satpy/readers/gerb_l2_hr_h5.py:
```diff
@@ -0,0 +1,99 @@
+        super(GERB_HR_FileHandler, self).__init__(filename,
+                                                  filename_info,
+                                                  filetype_info)
+        self._h5fh = h5py.File(self.filename, 'r')
```
This is no longer used I think. That should mean you can get rid of `__init__` completely.

djhoese commented on satpy/readers/gerb_l2_hr_h5.py:
```diff
@@ -0,0 +1,99 @@
+        if ds_name not in ['Solar Flux', 'Thermal Flux', 'Solar Radiance', 'Thermal Radiance']:
+            raise KeyError(f"{ds_name} is an unknown dataset for this reader.")
+
+        ds = gerb_get_dataset(self, f'Radiometry/{ds_name}', ds_info)
```
This could be rewritten to only pass `self["Radiometry..."]` and `ds_info`. I don't think you need `self` in that function otherwise, right?

pdebuyl commented:
I applied your recommendend simplification to the reader (get rid of ``__init__`` and bypass use of `hfile` in `gerb_get_dataset`.

Regarding the duplicate effort in the test, I don't know if it is worth it to do it.

djhoese reviewed (approved):
Looks good! Thanks.

pdebuyl commented:
Wiii, thanks @djhoese
